File: CRMAdapter/Vast/Adapter/AppointmentAdapter.vb
'-----------------------------------------------------------------------
' File: AppointmentAdapter.vb
' Role: Implements canonical appointment access for the VAST Desktop backend.
' Architectural Purpose: Normalizes service scheduling data into the CRM canonical appointment aggregate.
'-----------------------------------------------------------------------
Option Strict On
Option Explicit On

Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Data
Imports System.Data.Common
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks
Imports CRMAdapter.CommonConfig
Imports CRMAdapter.CommonContracts
Imports CRMAdapter.CommonDomain

Namespace CRMAdapter.Vast.Adapter
    ''' <summary>
    ''' Legacy VAST Desktop implementation of the <see cref="IAppointmentAdapter"/> contract.
    ''' </summary>
    Public NotInheritable Class AppointmentAdapter
        Inherits SqlAdapterBase
        Implements IAppointmentAdapter

        Private Const DefaultListLimit As Integer = 100

        Private Shared ReadOnly RequiredAppointmentKeys As String() = {
            "Appointment.__source",
            "Appointment.Id",
            "Appointment.CustomerId",
            "Appointment.VehicleId",
            "Appointment.Start",
            "Appointment.End",
            "Appointment.Advisor",
            "Appointment.Status",
            "Appointment.Location"
        }

        Private Shared ReadOnly AppointmentProjectionFields As String() = {
            "Id",
            "CustomerId",
            "VehicleId",
            "Start",
            "End",
            "Advisor",
            "Status",
            "Location"
        }

        Private ReadOnly _defaultListLimit As Integer

        ''' <summary>
        ''' Initializes a new instance of the <see cref="AppointmentAdapter"/> class.
        ''' </summary>
        Public Sub New(connection As DbConnection, fieldMap As FieldMap, Optional defaultListLimit As Integer = DefaultListLimit)
            MyBase.New(connection, fieldMap)
            _defaultListLimit = defaultListLimit
            MappingValidator.EnsureMappings(fieldMap, RequiredAppointmentKeys, NameOf(AppointmentAdapter))
        End Sub

        ''' <inheritdoc />
        Public Async Function GetByIdAsync(id As Guid, Optional cancellationToken As CancellationToken = Nothing) As Task(Of Appointment) Implements IAppointmentAdapter.GetByIdAsync
            Dim fieldMap = Me.FieldMap.GetTargets("Appointment", AppointmentProjectionFields)
            Dim source = Me.FieldMap.GetEntitySource("Appointment")
            Dim selectClause = String.Join(", ", fieldMap.Select(Function(kvp) $"{kvp.Value} AS [{kvp.Key}]"))
            Dim commandText = $"SELECT {selectClause} FROM {source} WHERE {fieldMap("Id")} = @id"

            Dim command = Await CreateCommandAsync(commandText, cancellationToken).ConfigureAwait(False)
            Using command
                AddParameter(command, "@id", id)
                Dim reader = Await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(False)
                Using reader
                    If Not Await reader.ReadAsync(cancellationToken).ConfigureAwait(False) Then
                        Return Nothing
                    End If

                    Return ReadAppointment(reader)
                End Using
            End Using
        End Function

        ''' <inheritdoc />
        Public Async Function GetByDateAsync([date] As DateTime, maxResults As Integer, Optional cancellationToken As CancellationToken = Nothing) As Task(Of IReadOnlyCollection(Of Appointment)) Implements IAppointmentAdapter.GetByDateAsync
            Dim limit = Math.Min(_defaultListLimit, If(maxResults > 0, maxResults, _defaultListLimit))
            Dim fieldMap = Me.FieldMap.GetTargets("Appointment", AppointmentProjectionFields)
            Dim source = Me.FieldMap.GetEntitySource("Appointment")
            Dim selectClause = String.Join(", ", fieldMap.Select(Function(kvp) $"{kvp.Value} AS [{kvp.Key}]"))
            Dim startField = fieldMap("Start")
            Dim startOfDay = [date].Date
            Dim endOfDay = startOfDay.AddDays(1)
            Dim commandText = $"SELECT TOP (@limit) {selectClause} FROM {source} WHERE {startField} >= @start AND {startField} < @end ORDER BY {startField}"

            Dim command = Await CreateCommandAsync(commandText, cancellationToken).ConfigureAwait(False)
            Using command
                AddParameter(command, "@limit", limit, DbType.Int32)
                AddParameter(command, "@start", startOfDay, DbType.DateTime2)
                AddParameter(command, "@end", endOfDay, DbType.DateTime2)
                Dim appointments As New List(Of Appointment)()
                Dim reader = Await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(False)
                Using reader
                    While Await reader.ReadAsync(cancellationToken).ConfigureAwait(False)
                        appointments.Add(ReadAppointment(reader))
                    End While
                End Using

                Return New ReadOnlyCollection(Of Appointment)(appointments)
            End Using
        End Function

        Private Shared Function ReadAppointment(reader As DbDataReader) As Appointment
            Return New Appointment(
                reader.GetGuid(reader.GetOrdinal("Id")),
                reader.GetGuid(reader.GetOrdinal("CustomerId")),
                reader.GetGuid(reader.GetOrdinal("VehicleId")),
                reader.GetDateTime(reader.GetOrdinal("Start")),
                reader.GetDateTime(reader.GetOrdinal("End")),
                reader.GetString(reader.GetOrdinal("Advisor")),
                reader.GetString(reader.GetOrdinal("Status")),
                reader.GetString(reader.GetOrdinal("Location")))
        End Function
    End Class
End Namespace

File: CRMAdapter/VastOnline/Adapter/AppointmentAdapter.cs
/*
 * File: AppointmentAdapter.cs
 * Purpose: Implements canonical appointment access for the Vast Online backend.
 * Security Considerations: Validates identifiers and time ranges, clamps limits, parameterizes SQL, and executes operations under centralized retry and rate limiting.
 * Example Usage: `var appointments = await adapter.GetByDateAsync(DateTime.UtcNow.Date, 100, cancellationToken);`
 */
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using CRMAdapter.CommonConfig;
using CRMAdapter.CommonContracts;
using CRMAdapter.CommonDomain;
using CRMAdapter.CommonInfrastructure;

namespace CRMAdapter.VastOnline.Adapter
{
    /// <summary>
    /// Vast Online implementation of the <see cref="IAppointmentAdapter"/> contract.
    /// </summary>
    public sealed class AppointmentAdapter : SqlAdapterBase, IAppointmentAdapter
    {
        private const int DefaultListLimit = 100;

        private static readonly string[] RequiredAppointmentKeys =
        {
            "Appointment.__source",
            "Appointment.Id",
            "Appointment.CustomerId",
            "Appointment.VehicleId",
            "Appointment.Start",
            "Appointment.End",
            "Appointment.Advisor",
            "Appointment.Status",
            "Appointment.Location"
        };

        private static readonly string[] AppointmentProjectionFields =
        {
            "Id",
            "CustomerId",
            "VehicleId",
            "Start",
            "End",
            "Advisor",
            "Status",
            "Location"
        };

        private readonly int _defaultListLimit;

        /// <summary>
        /// Initializes a new instance of the <see cref="AppointmentAdapter"/> class.
        /// </summary>
        /// <param name="connection">Database connection.</param>
        /// <param name="fieldMap">Field map configuration.</param>
        /// <param name="retryPolicy">Retry policy implementation.</param>
        /// <param name="logger">Logger for diagnostics.</param>
        /// <param name="rateLimiter">Rate limiter controlling throughput.</param>
        /// <param name="defaultListLimit">Optional default list limit.</param>
        public AppointmentAdapter(
            DbConnection connection,
            FieldMap fieldMap,
            ISqlRetryPolicy? retryPolicy = null,
            IAdapterLogger? logger = null,
            IAdapterRateLimiter? rateLimiter = null,
            int defaultListLimit = DefaultListLimit)
            : base(connection, fieldMap, retryPolicy, logger, rateLimiter)
        {
            if (defaultListLimit <= 0)
            {
                throw new ArgumentOutOfRangeException(nameof(defaultListLimit), "Default list limit must be positive.");
            }

            _defaultListLimit = defaultListLimit;
            MappingValidator.EnsureMappings(fieldMap, RequiredAppointmentKeys, nameof(AppointmentAdapter));
            MappingValidator.EnsureEntitySources(fieldMap, new[] { "Appointment" }, nameof(AppointmentAdapter));
        }

        /// <inheritdoc />
        public Task<Appointment?> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
        {
            if (id == Guid.Empty)
            {
                throw new InvalidAdapterRequestException("Appointment id must be provided.");
            }

            return ExecuteDbOperationAsync(
                "AppointmentAdapter.GetById",
                async ct =>
                {
                    var fieldMap = FieldMap.GetTargets("Appointment", AppointmentProjectionFields);
                    var source = FieldMap.GetEntitySource("Appointment");
                    var selectClause = string.Join(", ", fieldMap.Select(kvp => $"{kvp.Value} AS [{kvp.Key}]"));
                    var commandText = $"SELECT {selectClause} FROM {source} WHERE {fieldMap["Id"]} = @id";

                    await using var command = await CreateCommandAsync(commandText, ct).ConfigureAwait(false);
                    AddParameter(command, "@id", id);

                    await using var reader = await command.ExecuteReaderAsync(ct).ConfigureAwait(false);
                    if (!await reader.ReadAsync(ct).ConfigureAwait(false))
                    {
                        return null;
                    }

                    return ReadAppointment(reader);
                },
                cancellationToken);
        }

        /// <inheritdoc />
        public Task<IReadOnlyCollection<Appointment>> GetByDateAsync(
            DateTime date,
            int maxResults,
            CancellationToken cancellationToken = default)
        {
            var limit = EnforceLimit(maxResults, _defaultListLimit, nameof(maxResults));
            var startOfDay = date.Date;
            var endOfDay = startOfDay.AddDays(1);

            return ExecuteDbOperationAsync(
                "AppointmentAdapter.GetByDate",
                async ct =>
                {
                    var fieldMap = FieldMap.GetTargets("Appointment", AppointmentProjectionFields);
                    var source = FieldMap.GetEntitySource("Appointment");
                    var selectClause = string.Join(", ", fieldMap.Select(kvp => $"{kvp.Value} AS [{kvp.Key}]"));
                    var startField = fieldMap["Start"];
                    var commandText = $@"SELECT TOP (@limit) {selectClause}
FROM {source}
WHERE {startField} >= @start AND {startField} < @end
ORDER BY {startField};";

                    await using var command = await CreateCommandAsync(commandText, ct).ConfigureAwait(false);
                    AddParameter(command, "@limit", limit, DbType.Int32);
                    AddParameter(command, "@start", startOfDay, DbType.DateTime2);
                    AddParameter(command, "@end", endOfDay, DbType.DateTime2);

                    var appointments = new List<Appointment>();
                    await using var reader = await command.ExecuteReaderAsync(ct).ConfigureAwait(false);
                    while (await reader.ReadAsync(ct).ConfigureAwait(false))
                    {
                        appointments.Add(ReadAppointment(reader));
                    }

                    return (IReadOnlyCollection<Appointment>)new ReadOnlyCollection<Appointment>(appointments);
                },
                cancellationToken);
        }

        private static Appointment ReadAppointment(DbDataReader reader)
        {
            return new Appointment(
                reader.GetGuid(reader.GetOrdinal("Id")),
                reader.GetGuid(reader.GetOrdinal("CustomerId")),
                reader.GetGuid(reader.GetOrdinal("VehicleId")),
                reader.GetDateTime(reader.GetOrdinal("Start")),
                reader.GetDateTime(reader.GetOrdinal("End")),
                reader.GetString(reader.GetOrdinal("Advisor")),
                reader.GetString(reader.GetOrdinal("Status")),
                reader.GetString(reader.GetOrdinal("Location")));
        }
    }
}
